api/routes.py

```diff
@@ -126,9 +126,34 @@ def wrapped_function(*args, **kwargs):
         return resp
     return update_wrapper(wrapped_function, f)
 
+def allows_admin_web(f):
+    # Override Flask's default behavior and intercept the OPTIONS method for
+    # every request so CORS headers can be added.
+    f.required_methods = getattr(f, 'required_methods', set())
+    f.required_methods.add("OPTIONS")
+    f.provide_automatic_options = False
+
+    def wrapped_function(*args, **kwargs):
+        if request.method == "OPTIONS":
+            resp = app.make_default_options_response()
+        else:
+            resp = make_response(f(*args, **kwargs))
+
+        admin_web_domains = app.manager.admin_web_domains
+        if admin_web_domains:
+            options = get_cors_options(
+                app, dict(origins=admin_web_domains,
+                          supports_credentials=True)
+            )
+            set_cors_headers(resp, options)
+
+        return resp
+    return update_wrapper(wrapped_function, f)
```

Comment:
This function is a copy + paste of `allows_patron_web` function right above it... is there a way to not duplicate this code? I tried doing as Leonard suggested, where I rename the function to `allows_cors()` and declare two variables as so:

```
allows_patron_web = allows_cors(f, app.manager.patron_web_domains)
allows_admin_web = allows_cors(f, app.manager.admin_web_domains)
```
but that didn't seem to work when I used `allows_patron_web` or `allows_admin_web` as a decorator. i received errors that `f` was not defined.

Reply:
Yes, what I wrote was wrong. `f` is the function you're decorating and it's not available until later. Try this slight variation:

```
def allows_patron_web(f):
    return allows_cors(f, app.manager.patron_web_domains)

def allows_admin_web(f):
    return allows_cors(f, app.manager.admin_web_domains)
````

Reply:
Hmm this solved the `f is not defined` issue but now it is saying that it app doesn't have a manager property... do i need to wrap the function in order to get access to the app.manager?

Reply:
I don't know if this will get you all the way to a solution but I think it should help:

app.manager is set in api/routes.py in a function decorated with @app.before_first_request. For some reason, the allows_cors call is running before that function runs. It looks like your decorators are being called much earlier than I thought they would be.

Looking at the other decorator definitions in api/routes.py, I think the secret might be the @wraps decorator used _inside_ the definition of the main decorator. I think @wraps will make your decorators return a function which gets called much later, during an actual request -- by that time app.manager will be defined.

Reply:
Ok I pushed up a change to this decorator so that it can be re-used as `allows_cors(allowed_domain_type=set({"patron", "admin"}))`. @mwbenowitz let me know if this is what you had in mind. I think it is an improvement.